File: sdk-js/src/crypto/security_handshake.ts
/**
 * CODEX Fabric - Security Handshake
 * 
 * Orchestrates E2EE key exchange between peers.
 * Private keys NEVER leave the client device.
 */

import { KeyManager } from '../crypto/key_manager';
import { SignalingClient } from '../signaling/signaling_client';

export interface PeerKeys {
  peerId: string;
  signingPublicKey: string;
  exchangePublicKey: string;
}

export interface SessionInfo {
  peerId: string;
  peerSigningPublicKey: string;
  peerExchangePublicKey: string;
  localExchangePublicKey: string;
  establishedAt: Date;
}

export type HandshakeState = 'idle' | 'initializing' | 'ready' | 'exchanging' | 'established' | 'error';

/**
 * Manages the E2EE handshake protocol.
 * 
 * Flow:
 * 1. Each client generates key pairs locally
 * 2. Client A sends public keys to Client B via signaling server
 * 3. Client B receives, performs ECDH locally, derives session keys
 * 4. Client B sends acknowledgment back to Client A
 * 5. Client A also derives session keys independently
 * 6. Both peers now share identical session keys for AES-256-GCM encryption
 */
export class SecurityHandshake {
  private _keyManager: KeyManager;
  private _signaling: SignalingClient;
  private _state: HandshakeState = 'idle';
  private _peerKeys: Map<string, PeerKeys> = new Map();
  private _sessions: Map<string, SessionInfo> = new Map();

  onHandshakeComplete?: (peerId: string, session: SessionInfo) => void;
  onHandshakeFailed?: (peerId: string, error: string) => void;

  constructor(keyManager: KeyManager, signaling: SignalingClient) {
    this._keyManager = keyManager;
    this._signaling = signaling;
    this._setupListeners();
  }

  get state(): HandshakeState { return this._state; }

  async initialize(): Promise<void> {
    this._state = 'initializing';
    await this._keyManager.initialize();
    this._state = 'ready';
  }

  initiateKeyExchange(targetPeerId: string): void {
    if (!this._keyManager.isInitialized) throw new Error('Not initialized');
    this._state = 'exchanging';
    this._signaling.sendKeyExchange(
      targetPeerId,
      this._keyManager.signingPublicKey,
      this._keyManager.exchangePublicKey,
    );
  }

  async processKeyExchange(msg: {
    peer_id: string;
    signing_public_key: string;
    exchange_public_key: string;
  }): Promise<void> {
    if (!this._keyManager.isInitialized) throw new Error('Not initialized');

    this._peerKeys.set(msg.peer_id, {
      peerId: msg.peer_id,
      signingPublicKey: msg.signing_public_key,
      exchangePublicKey: msg.exchange_public_key,
    });

    await this._keyManager.deriveSessionKeys(msg.exchange_public_key);

    this._sessions.set(msg.peer_id, {
      peerId: msg.peer_id,
      peerSigningPublicKey: msg.signing_public_key,
      peerExchangePublicKey: msg.exchange_public_key,
      localExchangePublicKey: this._keyManager.exchangePublicKey,
      establishedAt: new Date(),
    });

    this._state = 'established';
    this._signaling.sendKeyExchangeAck(msg.peer_id, 'established');
    this.onHandshakeComplete?.call(this, msg.peer_id, this._sessions.get(msg.peer_id)!);
  }

  async processKeyExchangeAck(msg: { peer_id: string; status: string }): Promise<void> {
    if (msg.status !== 'established') {
      this._state = 'error';
      this.onHandshakeFailed?.call(this, msg.peer_id, 'Key exchange rejected');
      return;
    }

    const peerKeys = this._peerKeys.get(msg.peer_id);
    if (peerKeys) {
      await this._keyManager.deriveSessionKeys(peerKeys.exchangePublicKey);
      this._sessions.set(msg.peer_id, {
        peerId: msg.peer_id,
        peerSigningPublicKey: peerKeys.signingPublicKey,
        peerExchangePublicKey: peerKeys.exchangePublicKey,
        localExchangePublicKey: this._keyManager.exchangePublicKey,
        establishedAt: new Date(),
      });
      this._state = 'established';
      this.onHandshakeComplete?.call(this, msg.peer_id, this._sessions.get(msg.peer_id)!);
    }
  }

  encryptForPeer(plaintext: Buffer, peerId: string): Buffer {
    const peerKeys = this._peerKeys.get(peerId);
    if (!peerKeys) throw new Error(`No session with ${peerId}`);
    return this._keyManager.encrypt(plaintext, peerKeys.exchangePublicKey);
  }

  decryptFromPeer(ciphertext: Buffer, peerId: string): Buffer {
    const peerKeys = this._peerKeys.get(peerId);
    if (!peerKeys) throw new Error(`No session with ${peerId}`);
    return this._keyManager.decrypt(ciphertext, peerKeys.exchangePublicKey);
  }

  isSessionEstablished(peerId: string): boolean {
    return this._sessions.has(peerId);
  }

  clearSessions(): void {
    this._keyManager.clearSessionKeys();
    this._peerKeys.clear();
    this._sessions.clear();
    this._state = 'idle';
  }

  private _setupListeners(): void {
    this._signaling.on('key-exchange', (msg) => {
      this.processKeyExchange(msg as any);
    });
    this._signaling.on('key-exchange-ack', (msg) => {
      this.processKeyExchangeAck(msg as any);
    });
  }
}

File: sdk/lib/src/crypto/security_handshake.dart
import 'dart:typed_data';

import 'key_manager.dart';
import '../signaling/messages.dart';

/// Orchestrates the end-to-end encryption (E2EE) handshake between peers.
///
/// This manager coordinates the key exchange protocol ensuring that:
/// - Cryptographic keys are generated exclusively on the client device
/// - Private keys NEVER touch the signaling server or any intermediate node
/// - Only public keys are transmitted over the signaling channel
/// - Session encryption keys are derived locally using ECDH + HKDF
///
/// ## Handshake Flow
///
/// ```
/// Client A                                    Client B
///   |                                            |
///   |-- generateKeys()                           |-- generateKeys()
///   |                                            |
///   |--- key-exchange (pubKeys) ---------------->|
///   |                                            |-- deriveSessionKeys(A.pub)
///   |<-------- key-exchange (pubKeys) -----------|
///   |-- deriveSessionKeys(B.pub)                 |
///   |                                            |
///   |========= E2EE Channel Established =========|
/// ```
///
/// The signaling server acts only as a relay for public keys. It has
/// zero knowledge of the derived session keys and cannot decrypt traffic.
class SecurityHandshake {
  /// The local key manager handling cryptographic operations
  final KeyManager _keyManager;

  /// State machine tracking the handshake progress
  SecurityHandshakeState _state = SecurityHandshakeState.idle;

  /// Map of peer ID -> their public keys
  final Map<String, PeerPublicKeys> _peerKeys = {};

  /// Map of peer exchange public key -> session keys
  final Map<String, SessionInfo> _establishedSessions = {};

  /// Callback when handshake completes for a peer
  void Function(String peerId, SessionInfo session)? onHandshakeComplete;

  /// Callback when handshake fails
  void Function(String peerId, String error)? onHandshakeFailed;

  /// Create a new SecurityHandshake instance
  SecurityHandshake({
    required KeyManager keyManager,
    this.onHandshakeComplete,
    this.onHandshakeFailed,
  }) : _keyManager = keyManager;

  /// Get the current handshake state
  SecurityHandshakeState get state => _state;

  /// Get the local key manager
  KeyManager get keyManager => _keyManager;

  /// Whether E2EE is established with a specific peer
  bool isSessionEstablished(String peerId) {
    final peerKeys = _peerKeys[peerId];
    if (peerKeys == null) return false;
    return _establishedSessions.containsKey(peerKeys.exchangePublicKey);
  }

  /// Get session info for a peer (if established)
  SessionInfo? getSessionInfo(String peerId) {
    final peerKeys = _peerKeys[peerId];
    if (peerKeys == null) return null;
    return _establishedSessions[peerKeys.exchangePublicKey];
  }

  /// Initialize the handshake by initializing the key manager.
  ///
  /// This generates the signing and exchange key pairs locally.
  /// Call this once when the client connects to the signaling server.
  Future<void> initialize() async {
    _state = SecurityHandshakeState.initializing;
    
    try {
      await _keyManager.initialize();
      _state = SecurityHandshakeState.ready;
    } catch (e) {
      _state = SecurityHandshakeState.error;
      rethrow;
    }
  }

  /// Create a key exchange message to send to a specific peer.
  ///
  /// This message contains ONLY the local public keys. The private keys
  /// remain securely stored in the local KeyManager and are never serialized.
  ///
  /// Returns a [KeyExchangeMessage] ready to be sent via the signaling server.
  Future<KeyExchangeMessage> createKeyExchangeMessage(String targetPeerId) async {
    if (!_keyManager.isInitialized) {
      throw StateError('KeyManager not initialized. Call initialize() first.');
    }

    _state = SecurityHandshakeState.exchanging;

    return KeyExchangeMessage(
      peerId: targetPeerId,
      signingPublicKey: _keyManager.signingPublicKey,
      exchangePublicKey: _keyManager.exchangePublicKey,
    );
  }

  /// Process an incoming key exchange message from a remote peer.
  ///
  /// This method:
  /// 1. Stores the peer's public keys
  /// 2. Performs ECDH key exchange using our private key + their public key
  /// 3. Derives session encryption keys locally
  /// 4. Returns an acknowledgment message
  ///
  /// The server never sees our private key or the derived session keys.
  ///
  /// Returns a [KeyExchangeAckMessage] to send back to the peer.
  Future<KeyExchangeAckMessage> processKeyExchangeMessage(
    KeyExchangeMessage message,
  ) async {
    if (!_keyManager.isInitialized) {
      throw StateError('KeyManager not initialized. Call initialize() first.');
    }

    // Store the peer's public keys
    final peerKeys = PeerPublicKeys(
      peerId: message.peerId,
      signingPublicKey: message.signingPublicKey,
      exchangePublicKey: message.exchangePublicKey,
    );
    _peerKeys[message.peerId] = peerKeys;

    // Perform ECDH and derive session keys
    // This uses our private key + their public key to create shared secrets
    // The server cannot derive these keys because it never sees either private key
    await _keyManager.deriveSessionKeys(
      message.exchangePublicKey,
    );

    // Create session info
    final sessionInfo = SessionInfo(
      peerId: message.peerId,
      peerSigningPublicKey: message.signingPublicKey,
      peerExchangePublicKey: message.exchangePublicKey,
      localExchangePublicKey: _keyManager.exchangePublicKey,
      establishedAt: DateTime.now(),
    );
    _establishedSessions[message.exchangePublicKey] = sessionInfo;

    _state = SecurityHandshakeState.established;

    // Notify callback
    onHandshakeComplete?.call(message.peerId, sessionInfo);

    // Return acknowledgment containing our public keys so the initiating
    // peer can complete the handshake (server will relay this to the peer).
    return KeyExchangeAckMessage(
      peerId: message.peerId,
      status: 'established',
      signingPublicKey: _keyManager.signingPublicKey,
      exchangePublicKey: _keyManager.exchangePublicKey,
    );
  }

  /// Process a key exchange acknowledgment from a peer.
  ///
  /// This completes the handshake from the initiator's side, storing
  /// the peer's public keys and deriving session encryption keys.
  Future<void> processKeyExchangeAck(KeyExchangeAckMessage ack) async {
    if (!_keyManager.isInitialized) {
      throw StateError('KeyManager not initialized.');
    }

    if (ack.status != 'established') {
      onHandshakeFailed?.call(ack.peerId, 'Key exchange rejected: ${ack.status}');
      _state = SecurityHandshakeState.error;
      return;
    }

    // The acknowledgment carries the responder's public keys (added by the
    // server when relaying the ack back to us).
    final signingPublicKey = ack.signingPublicKey;
    final exchangePublicKey = ack.exchangePublicKey;
    if (signingPublicKey == null || exchangePublicKey == null) {
      throw StateError('Key exchange ack is missing peer public keys');
    }

    // Store the peer's public keys
    _peerKeys[ack.peerId] = PeerPublicKeys(
      peerId: ack.peerId,
      signingPublicKey: signingPublicKey,
      exchangePublicKey: exchangePublicKey,
    );

    // Derive session keys using the responder's exchange public key
    await _keyManager.deriveSessionKeys(exchangePublicKey);

    // Store the established session
    final sessionInfo = SessionInfo(
      peerId: ack.peerId,
      peerSigningPublicKey: signingPublicKey,
      peerExchangePublicKey: exchangePublicKey,
      localExchangePublicKey: _keyManager.exchangePublicKey,
      establishedAt: DateTime.now(),
    );
    _establishedSessions[exchangePublicKey] = sessionInfo;
    _state = SecurityHandshakeState.established;
    onHandshakeComplete?.call(ack.peerId, sessionInfo);
  }

  /// Encrypt data for a specific peer using the established session keys.
  ///
  /// This is a convenience method that delegates to the KeyManager's
  /// encryption using the session keys derived during the handshake.
  Future<Uint8List> encryptForPeer(Uint8List plaintext, String peerId) async {
    final peerKeys = _peerKeys[peerId];
    if (peerKeys == null) {
      throw StateError('No keys for peer $peerId. Complete handshake first.');
    }

    return _keyManager.encrypt(plaintext, peerKeys.exchangePublicKey);
  }

  /// Decrypt data from a specific peer using the established session keys.
  Future<Uint8List> decryptFromPeer(Uint8List ciphertext, String peerId) async {
    final peerKeys = _peerKeys[peerId];
    if (peerKeys == null) {
      throw StateError('No keys for peer $peerId. Complete handshake first.');
    }

    return _keyManager.decrypt(ciphertext, peerKeys.exchangePublicKey);
  }

  /// Sign data using the local Ed25519 signing key.
  ///
  /// The signature can be verified by any peer who has our signing public key.
  Future<Uint8List> signData(Uint8List data) async {
    return _keyManager.sign(data);
  }

  /// Verify a signature from a peer using their stored signing public key.
  Future<bool> verifyPeerSignature(
    Uint8List data,
    Uint8List signature,
    String peerId,
  ) async {
    final peerKeys = _peerKeys[peerId];
    if (peerKeys == null) {
      throw StateError('No keys for peer $peerId.');
    }

    return _keyManager.verify(data, signature, peerKeys.signingPublicKey);
  }

  /// Export all public keys for sharing via the signaling channel.
  ///
  /// This is safe to transmit - it contains only public information.
  Map<String, dynamic> exportPublicKeys() {
    return _keyManager.exportPublicKeys();
  }

  /// Clear all session data and peer keys.
  ///
  /// Call this when leaving a room or disconnecting.
  void clearSessions() {
    _keyManager.clearSessionKeys();
    _peerKeys.clear();
    _establishedSessions.clear();
    _state = SecurityHandshakeState.idle;
  }

  /// Dispose of all resources.
  void dispose() {
    clearSessions();
    _keyManager.dispose();
    _state = SecurityHandshakeState.disposed;
  }
}

/// States of the security handshake process
enum SecurityHandshakeState {
  /// No handshake in progress
  idle,

  /// Key manager is being initialized
  initializing,

  /// Keys are ready, waiting to start exchange
  ready,

  /// Key exchange is in progress with peers
  exchanging,

  /// E2EE session is established
  established,

  /// An error occurred during handshake
  error,

  /// The handshake has been disposed
  disposed,
}

/// Public keys from a remote peer
class PeerPublicKeys {
  /// The peer's ID
  final String peerId;

  /// Peer's Ed25519 signing public key (hex)
  final String signingPublicKey;

  /// Peer's X25519 exchange public key (hex)
  final String exchangePublicKey;

  PeerPublicKeys({
    required this.peerId,
    required this.signingPublicKey,
    required this.exchangePublicKey,
  });

  Map<String, dynamic> toJson() => {
    'peer_id': peerId,
    'signing_public_key': signingPublicKey,
    'exchange_public_key': exchangePublicKey,
  };

  factory PeerPublicKeys.fromJson(Map<String, dynamic> json) {
    return PeerPublicKeys(
      peerId: json['peer_id'] as String,
      signingPublicKey: json['signing_public_key'] as String,
      exchangePublicKey: json['exchange_public_key'] as String,
    );
  }
}

/// Information about an established E2EE session
class SessionInfo {
  /// The peer's ID
  final String peerId;

  /// Peer's Ed25519 signing public key (hex)
  final String peerSigningPublicKey;

  /// Peer's X25519 exchange public key (hex)
  final String peerExchangePublicKey;

  /// Our X25519 exchange public key (hex)
  final String localExchangePublicKey;

  /// When the session was established
  final DateTime establishedAt;

  SessionInfo({
    required this.peerId,
    required this.peerSigningPublicKey,
    required this.peerExchangePublicKey,
    required this.localExchangePublicKey,
    required this.establishedAt,
  });

  /// Duration since the session was established
  Duration get uptime => DateTime.now().difference(establishedAt);

  Map<String, dynamic> toJson() => {
    'peer_id': peerId,
    'peer_signing_public_key': peerSigningPublicKey,
    'peer_exchange_public_key': peerExchangePublicKey,
    'local_exchange_public_key': localExchangePublicKey,
    'established_at': establishedAt.toIso8601String(),
  };
}